Initialize contracts having function or event overrides

reedsa opened the pull request:
### What was wrong?

Closes #1704, #3482

The `Contract` class cannot initialize if the contract contains overloaded functions or events. Currently only the function/event name is used to filter ABI events from the contract ABI, so an exception is raised. Currently, events raise a ValueError when `_get_event_abi` is called on the Contract or ContractEvent class. Functions raise a MismatchedABI error.

Example:
`BaseContractEvent._get_event_abi()` calls eth-utils `get_event_abi()` and expects to return the ABI for the event class. The event classes are not created for every overloaded function. When there are contract events with the same name, only one is ever created and used.

### How was it fixed?

Initialize the `ContractFunctions` or `ContractEvents` class with every ABI found in the contract, storing the abi signature in the class name. Along with the signature, the `abi` property is set for each class within the `__init__` method. This ensures the correct instance is found and used for subsequent functions like `get_logs()`.

`get_abi_element` now handles both functions and events using the ABI signature to find a unique ABI. The signature is represented by a string that includes it's name and list of input types within parentheses. `name(arg1Type,arg2Type,...)`

Exceptions are now shared between functions and events for consistency.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://www.aramcoworld.com/getattachment/d93f40b7-c649-4db8-a79c-8b3156431f5b/Camels-The-Magnificent-Migration?width=1400)

Commit:
Ambiguous event contract test

Commit:
Use ``get_abi_element`` instead of ``get_event_abi``

Commit:
Failing test for ``get_abi_element`` for retrieving an overloaded event ABI

Commit:
Better filtering in ``get_abi_element``

Includes ``filter_by_argument_type``, ``get_name_by_abi_element_identifier``, ``_get_abi_by_signature``
Simplify ``_mismatched_abi_error_diagnosis`` arguments.

Commit:
Use argument types to disambiguate event ABIs in ``_get_event_abi`` calls

Commit:
Update ContractFunction methods to call ``_get_abi()`` and handle ambiguity

Commit:
Handle ambiguity to get the correct function ABI in ``prepare_transaction``

Commit:
Find ambiguous function without args in ``_find_matching_fn_abi``

Commit:
Use ABI from Contract Event inside class methods.

Commit:
Resolve ambiguous event references by storing a copy of each event with the event signature

Commit:
Use the event interface to get the correct instance

Commit:
Newsfragment for #3476

kclowes commented on tests/core/contracts/conftest.py:
```diff
@@ -760,3 +785,4 @@ def async_call(request):
 @pytest.fixture
 def async_estimate_gas(request):
     return async_partial(async_invoke_contract, api_call_desig="estimate_gas")
+    return async_partial(async_invoke_contract, api_call_desig="estimate_gas")
```
copy pasta?

kclowes commented on tests/core/contracts/test_contract_ambiguous_events.py:
```diff
@@ -0,0 +1,260 @@
+    assert event_abi == ABI_EVENT_DEPOSIT
+
+
+def test_get_abi_element_with_signature_for_amibguous_tuple_events() -> None:
```
nit: 
```suggestion
def test_get_abi_element_with_signature_for_amibiguous_tuple_events() -> None:
```

kclowes commented on tests/core/utilities/test_abi.py:
```diff
@@ -510,14 +535,6 @@ def test_get_abi_element(
             "Unsupported function identifier",
         ),
-        (
-            CONTRACT_ABI,
```
Curious about this removal?

reedsa replied:
Meant to move this to the `test_get_abi_element`, since now `get_abi_element` will determine the function by matching it's name (without arguments). Since `logTwoEvents` is the only function with that name, this case works.

I'll add a new failing case for an ambiguous function name that raises a MismatchedABI.

kclowes commented on web3/contract/base_contract.py:
```diff
@@ -148,19 +152,27 @@ class BaseContractEvent:
+    def __init__(
+        self,
+        *argument_names: Tuple[str],
+        abi: ABIEvent = None,
```
nit?  
```suggestion
        abi: Optional[ABIEvent] = None,
```

kclowes commented on web3/contract/base_contract.py:
```diff
@@ -422,30 +432,40 @@ def __init__(
                 "Are you sure you provided the correct contract abi?",
             )
-        elif event_name not in self.__dict__["_events"]:
+        elif event_name.split("(")[0] not in [
```
this `event_name.split...` may deserve to be its own method since you're using it so much in here. Feel free to take or leave :)

kclowes commented on newsfragments/3476.feature.rst:
```diff
@@ -0,0 +1 @@
+Contracts which overloaded functions or events are now supported. The Contract initializes functions and events using an identifier to distinguish between them. The identifier is the function or event signature, which consists of the name and the parameter types.
```
nit:
```suggestion
Contracts with overloaded functions or events are now supported. The Contract initializes functions and events using an identifier to distinguish between them. The identifier is the function or event signature, which consists of the name and the parameter types.
```

kclowes commented on web3/contract/base_contract.py:
```diff
@@ -490,19 +510,39 @@ def __init__(self, abi: Optional[ABIFunction] = None) -> None:
-                    *self.args,
-                    abi_codec=self.w3.codec,
-                    **self.kwargs,
+                    cls.contract_abi, get_abi_element_identifier(abi_element_identifier)
```
Does this need the codec passed too?

kclowes reviewed (approved):
LGTM! Just left a few nits!

reedsa commented:
The ambiguity with functions is also an issue unfortunately. Taking a look at how to disambiguate since right now it uses the "first" one.

Commit:
Updates based on feedback

Commit:
Get function by argument type to resolve ambiguities

Commit:
All function classes initialize with the ABI

__getattr__ now uses the ABI to look up the function class by signature
Additional fixes for ens

Commit:
Additional test for ``get_function_by_signature``

Refactoring and docs update for ``get_abi_element`` and ``_build_filters``

Rename ``get_abi_element_identifier`` -> ``get_abi_element_signature``

Commit:
Deprecate ``get_event_abi``

Commit:
Rename ``_find_abi_identifier_by_name`` -> ``_get_any_abi_signature_with_name``

Further cleanup and refactoring

Commit:
Compile with latest Solidity version (v0.8.27)

Commit:
Add test case for ``abi`` used as function name in a contract

Fix contract compilation and naming

pacrob commented on tests/ens/test_offchain_resolution.py:
```diff
@@ -173,14 +173,14 @@ def test_offchain_resolver_function_call_raises_with_ccip_read_disabled(
         offchain_resolver.functions.resolve(
             ens_encode_name("offchainexample.eth"),
-            ENCODED_ADDR_CALLDATA,
+            ENCODED_ADDR_CALLDATA.encode(),
```
Why the added `encode` here? Test passes for me with or without it.

reedsa replied:
No need for this after all, good callout.

pacrob commented on web3/contract/contract.py:
```diff
@@ -245,28 +255,55 @@ def __init__(
+
+    def __iter__(self) -> Iterable["ContractEvent"]:
+        """
+        Iterate over supported
```
nit - docstring should state what the function does (not sure if it's necessary for a magic method?). Also should match for the same method in `AsyncContractEvents`.

reedsa replied:
Good point, there's no practical information I provided there anyway. Removing!

pacrob commented on web3/utils/abi.py:
```diff
@@ -306,24 +449,27 @@ def get_abi_element_info(
     """
-    Return the interface for an ``ABIElement`` which matches the provided identifier
-    and arguments.
+    Return the interface for an ``ABIElement`` from the `abi` that matches the provided
```
```suggestion
    Return the interface for an ``ABIElement`` from the ``abi`` that matches the provided
```

pacrob commented on web3/utils/abi.py:
```diff
@@ -472,12 +598,17 @@ def check_if_arguments_can_be_encoded(
     """
+    .. warning::
+        This function is deprecated. It is unable to distinguish between
+        overloaded events. Use `get_abi_element` instead.
```
```suggestion
        overloaded events. Use ``get_abi_element`` instead.
```

Commit:
Update event_contracts.py

Commit:
Revert changes made to ens test

Remove docs from events iter
Fix docstring formatting

Commit:
Add section to ambiguous section of the contract docs

pacrob commented on web3/utils/abi.py:
```diff
@@ -231,6 +274,106 @@ def _get_argument_readable_type(arg: Any) -> str:
+    filter is a partial function that takes a contract ABI and returns a filtered list.
+    Each parameter is checked before applying the relevant filter.
+
+    When the `abi_element_identifier` is a function name or signature and no arguments
```
```suggestion
    When the ``abi_element_identifier`` is a function name or signature and no arguments
```

pacrob commented on web3/utils/abi.py:
```diff
@@ -231,6 +274,106 @@ def _get_argument_readable_type(arg: Any) -> str:
+    filters include the function name, argument count, argument name, argument type,
+    and argument encodability.
+
+    `constructor`, `fallback`, and `receive` ABI elements are handled only with a
```
```suggestion
    ``constructor``, ``fallback``, and ``receive`` ABI elements are handled only with a
```

pacrob commented on web3/utils/abi.py:
```diff
@@ -306,24 +449,27 @@ def get_abi_element_info(
-    :param abi_element_identifier: Find an element ABI with matching identifier.
+    :param abi_element_identifier: Find an element ABI with matching identifier. The \
+    identifier may be a function name, signature, or ``FallbackFn`` or ``ReceiveFn``. \
+    A function signature is in the form `name(arg1Type,arg2Type,...)`.
```
```suggestion
    A function signature is in the form ``name(arg1Type,arg2Type,...)``.
```

Commit:
Improved docs formatting

pacrob reviewed (approved):
:sheep:

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+    function_name_tester_contract_abi,
+    function_name_tester_contract,
+):
+    # test `w3` function name does not throw when creating the contract factory
```
```suggestion
    # test `abi` function name does not throw when creating the contract factory
```

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+    # re-instantiate the contract
+    contract = contract_factory(function_name_tester_contract.address)
+
+    # Contract w3 function should not override web3 instance
```
Maybe something to test that `contract.abi` doesn't get overridden here instead?

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+
+    assert contract.abi == function_name_tester_contract_abi
+
+    # assert the `w3` function returns true when called
```
```suggestion
    # assert the `abi` function returns true when called
```

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+    function_name_tester_contract_abi,
+    async_function_name_tester_contract,
+):
+    # test `w3` function name does not throw when creating the contract factory
```
```suggestion
    # test `abi` function name does not throw when creating the contract factory
```

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+
+    # Contract w3 function should not override web3 instance
+    with pytest.raises(AttributeError):
+        contract.functions.abi.get_block("latest")
```
Maybe a test to make sure `abi` doesn't get overridden here instead?

reedsa replied:
I made a slight adjustment to the order of the assertion/exception cases and updated the inline comments to be accurate. Let me know if there's further clarity needed there!

kclowes commented on tests/core/contracts/test_contract_class_construction.py:
```diff
@@ -43,6 +43,51 @@ def test_abi_as_json_string(w3, math_contract_abi, some_address):
+
+    assert contract.abi == function_name_tester_contract_abi
+
+    # assert the `w3` function returns true when called
```
```suggestion
    # assert the `abi` function returns true when called
```

kclowes commented on web3/contract/base_contract.py:
```diff
@@ -486,23 +516,49 @@ class BaseContractFunction:
     kwargs: Any = None
 
-    def __init__(self, abi: Optional[ABIFunction] = None) -> None:
+    def __init__(self, *arguments: Any, abi: Optional[ABIFunction] = None) -> None:
```
This is a breaking change, since I believe now abi has to be passed in as a kwarg. Does that change anything for the UI?

reedsa replied:
This was originally intended to bring the same interface to `BaseContractFunction` and `BaseContractEvent`, but the `arugments` are never actually used.

I've also taken further steps to eliminate extraneous arguments passed to the factory function. Passing the signature is enough to find the ABI that is needed later on. I've made updates to the async/sync contracts so initialization of the ABI still works as intended.

kclowes commented on web3/contract/base_contract.py:
```diff
@@ -148,19 +154,55 @@ class BaseContractEvent:
+        abi: Optional[ABIEvent] = None,
+        abi_element_identifier: Optional[ABIElementIdentifier] = None,
+    ) -> None:
+        # event was referenced with parens, juts return the event
```
nit:
```suggestion
        # event was referenced with parens, just return the event
```

kclowes commented on web3/contract/utils.py:
```diff
@@ -119,6 +122,34 @@ def format_contract_call_return_data_curried(
     return normalized_data[0] if len(normalized_data) == 1 else normalized_data
 
 
+def copy_contract_function(
```
I'm not sure we gain anything by exposing this publicly. Consider moving these two methods to `web3/_utils/contracts.py`

reedsa replied:
Good point, moved!

kclowes commented on tests/core/contracts/test_contract_ambiguous_events.py:
```diff
@@ -0,0 +1,260 @@
+
+    event_get_abi_result = ambiguous_event_contract.events[
+        "LogSingleArg(uint256)"
+    ]._get_event_abi()
```
This is a stylistic nitpick, but it seems a little smelly to me to test a private method. It looks like `_get_event_abi` is used to assign the abi attribute on `ContractEvent`. Would it be possible to do something like: 

```suggestion
event_get_abi_result = ambiguous_event_contract.events[
    "LogSingleArg(uint256)"
].abi
```

Feel free to take or leave if that doesn't make sense.

reedsa replied:
This is definitely better as `.abi` 👍

kclowes reviewed (commented):
Sweet! I didn't see anything major, just left a few small comments.

Commit:
Add ``@combomethod`` to ``_parse_logs`` in base_contract

Commit:
Initialize and use correct ``fallback`` or ``receive`` function

Filter by type prior to getting an ABI element

Commit:
Apply changes from #3479

Commit:
Feedback improvements

pacrob reviewed (approved):
:sheep:

kclowes reviewed (approved):
LGTM! Since this is such a core component of our library, it's probably worth some manual testing to make sure old functionality still works as expected. I did some myself and have fairly high confidence in the tests, but it's probably worth one more run through!

Commit:
Clean up logic in filters

Commit:
Fix logic
